compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -832,6 +844,43 @@ object desugar {
         companionDefs(anyRef, companionMembers)
       else if isValueClass && !isObject then
         companionDefs(anyRef, Nil)
+      else if (isJavaRecord) {
+
+        /** Get the canonical constructor of the Java record.
+          *
+          * Java classes have a dummy constructor; see [[JavaParsers.makeTemplate]] for
+          * more details
+          */
+        def canonicalConstructor(impl: Template): DefDef = {
+          impl.body.collectFirst {
+            case ddef: DefDef if ddef.name.isConstructorName && ddef.mods.is(Synthetic) =>
+              ddef
+          }.get
+        }
+
+        val constr1 = canonicalConstructor(impl)
```

Comment:
Need to tidy some of this up.

Reply:
Reminder:

1. Check whether we go through this properly when reading a jar.
2. Check whether scalac does something special in the gen b code phase for "companion methods" for java defined classes, or we literally create forwarders in the actual companion module?